File: output/invoicing/currencyBusiness.ts

//import { Currency, CurrencyCreate, CurrencyUpdate, CurrencyPartial, CurrencyView } from "./currencyClasses"
import { ICurrencyCreate, ICurrencyUpdate, ICurrencyPartial, ICurrencyView } from "./currencyInterfaces";
import { IQueryResult, IQuery, Context, Business } from "./base";

export class CurrencyBusiness extends Business<ICurrencyView> {

    constructor(context:Context) {
        super(context, "currency");
    }

    override createProperties: any = {
  "id": {
    "required": true,
    "type": "str"
  },
  "name": {
    "required": true,
    "type": "str"
  },
  "symbol": {
    "required": true,
    "type": "str"
  }
};
    override updateProperties: any = {
  "id": {
    "required": true,
    "type": "str"
  },
  "name": {
    "required": true,
    "type": "str"
  },
  "symbol": {
    "required": true,
    "type": "str"
  }
};
    override partialProperties: any = {
  "id": {
    "required": false,
    "type": "str"
  },
  "name": {
    "required": false,
    "type": "str"
  },
  "symbol": {
    "required": false,
    "type": "str"
  }
};
    
    override getAll = async ():Promise<IQueryResult<IQuery, ICurrencyView>> => {
        return super.getAll() as Promise<IQueryResult<IQuery, ICurrencyView>>;
    };

    override create = async (currency:ICurrencyCreate):Promise<ICurrencyView> => {
        return super.create(currency) as Promise<ICurrencyView>;
    };

    override getById = async (id:string):Promise<ICurrencyView> => {
        return super.getById(id) as any;    
    };

    override update = async (id:string, currency:ICurrencyUpdate):Promise<ICurrencyView> => {
        return super.update(id, currency) as Promise<ICurrencyView>;
    };

    override modify = async (id:string, currency:ICurrencyPartial):Promise<ICurrencyView> => {
        return super.modify(id, currency) as Promise<ICurrencyView>;    
    };

    override delete = async (id:string):Promise<ICurrencyView> => {
        return super.delete(id) as Promise<ICurrencyView>;
    }; 
}

File: output/invoicing/Business/CurrencyBusiness.cs
using System.Linq.Expressions;
using System.Text.Json;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using Microsoft.EntityFrameworkCore.Storage;

public class CurrencyBusiness : Business<Currency, CurrencyView, CurrencyUpdate, CurrencyModify, CurrencyCreate, CurrencyQuery>
{
    public CurrencyBusiness(DbContext db) : base(db)
    {
    }

    public override DataQuery ConvertToDataQuery(CurrencyQuery query)
    {
        var dataQuery = base.ConvertToDataQuery(query);

        

        return dataQuery;
    }

    public override CurrencyQuery ConvertToClientQuery(DataQuery query)
    {
        var clientQuery = base.ConvertToClientQuery(query);

        foreach(var c in query.Where)
        {
            
        }        

        return clientQuery;
    }
    
    public override CurrencyView GetById(Guid id, int maxDepth = 2)
    {
        var query = Db.Set<Currency>()
            .Select(x => new CurrencyView { 
                Id = x.Id, Name = x.Name, Symbol = x.Symbol  
            })
            .AsQueryable();

        if (maxDepth > 0)
        {
            maxDepth--;
        }

        var entity = query.FirstOrDefault(x => x.Id == id) ?? throw new KeyNotFoundException($"No {entityName} entity found for given {id}");
        return entity;
    }

    public override CurrencyView Create(CurrencyCreate entity)
    {
        var dbSet = Db.Set<Currency>();
        var dbEntity = new Currency {
            Id = new Guid(),
            Name = entity.Name, Symbol = entity.Symbol
        };
        dbSet.Add(dbEntity);
        Db.SaveChanges();
        var added = dbSet.Select(x => new CurrencyView { 
                __EntityViewProjection__
            })
            .FirstOrDefault(x => x.Id == dbEntity.Id);
        
        if (added is null)
        {
            throw new Exception($"Could not retrieve created {entityName} entity.");
        }
        return added;
    }

    public override CurrencyView Update(Guid id, CurrencyUpdate entity)
    {
        var dbSet = Db.Set<Currency>();
        var existing = dbSet.Find(id);
        if (existing is null)
        {
            throw new KeyNotFoundException($"Could not find an existing {entityName} entity with the given id.");
        }

        var inputProps = typeof(CurrencyUpdate).GetProperties();
        var outputProps = typeof(Currency).GetProperties();

        foreach (var prop in inputProps)
        {
            if (prop.Name == "Id") continue;
            var match = outputProps.FirstOrDefault(p => p.Name == prop.Name);
            if (match is not null)
            {
                match.SetValue(existing, prop.GetValue(entity));
            }
        }

        Db.SaveChanges();
        var updated = GetById(id);
        return updated;
    }

    public override CurrencyView Modify(Guid id, JsonElement entity)
    {
        var dbSet = Db.Set<Currency>();
        var existing = dbSet.Find(id);
        if (existing is null)
        {
            throw new KeyNotFoundException($"Could not find an existing {entityName} entity with the given id.");
        }
      
        var validProps = typeof(CurrencyModify).GetProperties();
        var outputProps = typeof(Currency).GetProperties();

        foreach (JsonProperty prop in entity.EnumerateObject())
        {
            if (prop.Name.ToLower() == "id") continue;
            var match = outputProps.FirstOrDefault(p => p.Name.ToLower() == prop.Name.ToLower());
            if (match is not null)
            {
                match.SetValue(existing, prop.Value.GetString());//TODO: proper mapping of type
            }
        }

        Db.SaveChanges();
        var updated = GetById(id);
        return updated;
    }

    public override CurrencyView Delete(Guid id)
    {
        var dbSet = Db.Set<Currency>();
        var existing = dbSet.Find(id);
        if (existing is null)
        {
            throw new KeyNotFoundException($"Could not find an existing {entityName} entity with the given id.");
        }
        var beforeDelete = GetById(id);
        dbSet.Remove(existing);
        Db.SaveChanges();

        return beforeDelete;
    }

    public override QueryResult<ClientQuery, CurrencyView> GetAll(CurrencyQuery clientQuery, DataQuery query, int maxDepth = 2)
    {
        var q = Db.Set<Currency>().Skip(query.Offset);
                           
        if ( query.Limit > 0) 
        {
            q = q.Take(query.Limit);
        }    

        if ( query.Where.Count > 0 )
        {
            foreach (var c in query.Where)
            {   
                __EntityWhereConditions__                   
            }
        }

        IOrderedQueryable<Currency>? sortedQ = null;
        if (query.Sort.Count > 0)
        {
            foreach (var s in query.Sort)
            {
                
                if (s.Column == "Name")
                {
                    sortedQ = s.Direction == SortDirection.Asc ? 
                        sortedQ is null ? q.OrderBy(x => x.Name) : sortedQ.ThenBy(x => x.Name) 
                        : sortedQ is null ? q.OrderByDescending( x => x.Name) : sortedQ.ThenByDescending(x => x.Name);
                }
                
            }
        }
        
        var data = (sortedQ ?? q)
            .Select(x => new CurrencyView { __EntityViewProjection__ })
            .ToList();

        var result = new QueryResult<ClientQuery, CurrencyView>(clientQuery)
        {
            Count = data.Count,
            Result = data,
            Total = data.Count
        };

        return result;
    }

}

